static/js/services/xml-forms.js

```diff
@@ -0,0 +1,158 @@
+var _ = require('underscore');
+
+(function () {
+
+  'use strict';
+
+  var inboxServices = angular.module('inboxServices');
+
+  /**
+   * Util functions available to a form doc's `.context` function for checking if
+   * a form is relevant to a specific contact.
+   */
+  var CONTEXT_UTILS = {
+    ageInYears: function(c) {
+      if (!c.date_of_birth) {
+        return;
+      }
+      var birthday = new Date(c.date_of_birth),
+          today = new Date();
+      return (today.getFullYear() - birthday.getFullYear()) +
+          (today.getMonth() < birthday.getMonth() ? -1 : 0) +
+          (today.getMonth() === birthday.getMonth() &&
+              today.getDate() < birthday.getDate() ? -1 : 0);
+    },
+  };
+
+  inboxServices.factory('XmlForms', [
+    '$q', '$parse', '$log', 'DB', 'Changes', 'Auth', 'UserContact', 'PLACE_TYPES',
+    function($q, $parse, $log, DB, Changes, Auth, UserContact, PLACE_TYPES) {
+
+      var listeners = {};
+
+      var getForms = function() {
+        return DB.get()
+          .query('medic/forms', { include_docs: true })
+          .then(function(res) {
+            var forms = res.rows
+              .filter(function(row) {
+                return row.doc._attachments.xml;
+              })
+              .map(function(row) {
+                return row.doc;
+              });
+            return $q.resolve(forms);
+          });
+      };
+
+      var init = getForms();
+
+      var evaluateExpression = function(expression, context, user) {
+        return $parse(expression)(CONTEXT_UTILS, { contact: context.doc, user: user });
+      };
+
+      var filterAll = function(forms, context, user) {
+        var promises = _.map(forms, _.partial(filter, _, context, user));
+        return $q.all(promises)
+          .then(function(resolutions) {
+            // always splice in reverse...
```

Comment:
Would writing this in a more functional style where you can't make this kind of ordering mistake help? I'm not too familiar with underscore, but something like his:
```javascript
return $q.all(promises)
  .then(function(resolutions) {
    return $q.resolve(
      _.unzip(
        _.zip(forms, resolutions).filter(function(formAndRes) {
          return formAndRes[1];
        })
      )[0]);
  });
```

or

```javascript
return $q.all(promises)
  .then(function(resolutions) {
    return $q.resolve(_.reduce(_.zip(forms, resolutions), function(memo, formAndRes) {
      if (formAndRes[1]) {
        memo.push(formAndRes[0]);
      }
    }, []);
  });
```

There are probably cleaner ways from people smarter than me. I believe as well you'd no longer need to `forms.slice();` to clone them.